### PR TITLE
README.md: Add instructions for configuring debugger

### PR DESCRIPTION
**Pull request by grantcurell**
- Added detailed instructions for configuring a debugger to work on the project

Signed-off-by: Grant Curell <grant_curell@dell.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,6 +46,39 @@ You can refer our [Code of Conduct](https://github.com/dell/dellemc-openmanage-a
...
+    1. The location may be different for you but the main thing is `openmanage` must be accessible within the `ansible_collections.dellemc` namespace. That is to say, your path should look like `<SOME_PREFIX (usually dist-packages)>/ansible_collections/dellemc/openmanage`
+3. You will know it worked if things like `from ansible_collections.dellemc.openmanage.plugins.module_utils.dellemc_idrac import iDRACConnection` import cleanly and without error. Some IDEs may require you to restart the IDE in order to rescan the available packages.
+4. **NOTE:** In my IDE, `from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError` still reported an error. This error can be ignored safely. The code should still run correctly. Ansible is doing some import magic to make this work that may confuse your IDE's ability to check for the import correctly but it does actually run.
+5. One thing that isn't immediately obvious about debugging Ansible modules is how you inject arguments. This is described [in the Ansible docs](https://docs.ansible.com/ansible/2.9/dev_guide/developing_modules_general.html#exercising-module-code-locally).
```
Comment (anupamaloke) [inline README.md:56]: It would be good to point to the latest URL https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#exercising-module-code-locally

Comment (grantcurell) [reply README.md]: @anupamaloke updated!

**Review by jagadeeshnv — CHANGES_REQUESTED** [review]
@grantcurell 
Please add this content in the following location
https://github.com/dell/dellemc-openmanage-ansible-modules/tree/collections/.github/
".github/**debugging/README.md**"

**Comment by grantcurell** [comment]
@jagadeeshnv done!

**Review by jagadeeshnv — CHANGES_REQUESTED** [review]
@grantcurell **Please replace the worg 'calvin' with 'password'**. Thanks for this useful Debug info. Its more of "Direct speech" here. Please avoid using the words - I, my etc. Once this is done we will Pull it and get it reviewed by our Documentation team.

**Comment by grantcurell** [comment]
@jagadeeshnv - finished

**Comment by rajeevarakkal** [comment]
@grantcurell  - looks fine with me, all CI/CD checks also passing

@jagadeeshnv  -  scan once with SCA and semantics validation, in case any observations lets get that addressed.

**Comment by grantcurell** [comment]
@jagadeeshnv that is not for debugging in an IDE. It looks like that's for debugging a module post build. Like you are a user and want to know what's wrong. Step 4 looks like it results in the module source and step 5 is just telling you how to run it with Python.

**Comment by grantcurell** [comment]
My instructions tell you how to actually set our code up for debugging and how to step through it with a Python debugger. This is complicated by the fact that we have some non-standard library dependencies.

**Comment by jagadeeshnv** [comment]
@grantcurell So as i read it, The second part is more of Debugging using Pycharm(on Windows) via SSH interpreter on a linux machine and arguments json on the windows machine?

**Comment by grantcurell** [comment]
I wrote it for PyCharm step by step but that concept applies generally to any IDE. I've been messaging with someone else who lifted those instructions I wrote and used them for vscode. As mentioned in the instructions, you can use a remote Python session if you're on windows just to avoid having to get Ansible working on Windows (which is painful) or you can run it locally. The instructions explain both. I can demo it for you on a debugger of your choosing if it would help

**Comment by jagadeeshnv** [comment]
> Hi @grantcurell ,
> Can you try out this debugging https://docs.ansible.com/ansible/latest/dev_guide/debugging.html#detailed-debugging-steps
> is working with collection modules and is simpler than this. Any specific reason you doing this way?

The standard debugging instructions from https://docs.ansible.com/ansible/latest/dev_guide/debugging.html#detailed-debugging-steps results in `python AnsiballZ_ping.py execute` where we can use the debugger, in step 5 
The Method suggested by you result in  `python -m ansible.modules.some_module /tmp/args.json`. In both cases we pdb can be used to debug. Let us know if any specific advantage with your method?

File: README.md [hunk 2]
```diff
@@ -46,6 +46,10 @@ You can refer our [Code of Conduct](https://github.com/dell/dellemc-openmanage-a
 ## Testing
 See [here](https://github.com/dell/dellemc-openmanage-ansible-modules/tree/collections/tests/README.md) for further information on testing.
 
+## Debugging
+
+See [How to Debug](.github/debug.md)
```
Comment (rajeevarakkal) [inline README.md:51]: probably you can reword it to See how to debug OpenManage Ansible Modules using IDE
```suggestion
To debug OpenManage Ansible Modules using IDE, see [here](.github/debug.md)
```

File: .github/debug.md [hunk 3]
```diff
@@ -0,0 +1,47 @@
+# How to Perform Debugging
+
+The below instructions are written for Linux but include a method for using a remote debugger on Windows. Unless otherwise specified, the instructions are meant to be executed on a Linux machine.
```
Comment (rajeevarakkal) [inline .github/debug.md:3]:
```suggestion
The following enables you to debug OpenManage Ansible Modules from an IDE either from local Linux or from a remote debugger on Windows.
```

File: .github/debug.md [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+1. Clone the code to a directory of your choosing with your favorite git client or `git clone https://github.com/dell/dellemc-openmanage-ansible-modules.git`
```
Comment (rajeevarakkal) [inline .github/debug.md:5]: preferably you an remove this step, we are not seeing this cloned repository being used anywhere 
```suggestion

```

File: .github/debug.md [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+2. Download the code separately with `ansible-galaxy collection install dellemc.openmanage`. On Ubuntu this defaults to `/home/<YOUR_NAME>/.ansible/collections/ansible_collections/dellemc/openmanage/` The problem with this is that this location is not in PYTHONPATH which will cause problems with the debugger.
```
Comment (rajeevarakkal) [inline .github/debug.md:6]: Start this paragraph like `Install OpenManage Ansible Modules from Galaxy `

File: .github/debug.md [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+3. To resolve the pathing issues move the openmanage collection to align with the rest of Dell's code which is in PYTHONPATH with `sudo mv /home/<YOUR_NAME>/.ansible/collections/ansible_collections/dellemc/openmanage/ /usr/local/lib/python3.X/dist-packages/ansible_collections/dellemc/`. The path may be different on your system, but it should be placed with your other packages. Alternatively, you can add the directory `/home/<YOUR_NAME>/.ansible/collections/ansible_collections/dellemc/openmanage/` to PYTHONPATH with `export PYTHONPATH=$PYTHONPATH:/home/<YOUR_NAME>/.ansible/collections`. How to do this on Windows with a remote debugger is shown below.
```
Comment (rajeevarakkal) [inline .github/debug.md:7]: Would prefer to remove this `How to do this on Windows with a remote debugger is shown below.`. if needed you can explain this step in windows debugger section
```suggestion
3. To resolve the pathing issues move the openmanage collection to align with the rest of Dell's code which is in PYTHONPATH with `sudo mv /home/<YOUR_NAME>/.ansible/collections/ansible_collections/dellemc/openmanage/ /usr/local/lib/python3.X/dist-packages/ansible_collections/dellemc/`. The path may be different on your system, but it should be placed with your other packages. Alternatively, you can add the directory `/home/<YOUR_NAME>/.ansible/collections/ansible_collections/dellemc/openmanage/` to PYTHONPATH with `export PYTHONPATH=$PYTHONPATH:/home/<YOUR_NAME>/.ansible/collections`. 
```

File: .github/debug.md [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    1. The location may be different for but the key is `openmanage` must be accessible within the `ansible_collections.dellemc` namespace. That is to say, the path should look like `<SOME_PREFIX (usually dist-packages)>/ansible_collections/dellemc/openmanage`
+    2. When we tested, `from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError` still reported an error. This error can be ignored safely. Ansible is doing some import magic to make this work that confuses the IDE's ability to check for the import correctly but it does run.
```
Comment (rajeevarakkal) [inline .github/debug.md:9]:
```suggestion
    2. Sometimes, `from ansible.module_utils.six.moves.urllib.error import URLError, HTTPError` still reports an error. This error can be ignored safely. 
```

File: .github/debug.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    3. One thing that isn't immediately obvious about debugging Ansible modules is how to inject arguments. This is described [in the Ansible docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#exercising-module-code-locally).
```
Comment (rajeevarakkal) [inline .github/debug.md:10]: this additional information can be provide along with step 4, where we prepare args.json and inject it. Please this from here
```suggestion
    
```

File: .github/debug.md [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    4. If it worked, imports like `from ansible_collections.dellemc.openmanage.plugins.module_utils.dellemc_idrac import iDRACConnection` should import cleanly and without error. Some IDEs may require a restart in order to rescan the available packages.
```
Comment (rajeevarakkal) [inline .github/debug.md:11]:
```suggestion
    4. Some IDEs may require a restart in order to rescan the available packages for import statements to resolve.
```

File: .github/debug.md [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+            "ANSIBLE_MODULE_ARGS": {
+                "idrac_ip": "192.168.1.63",
+                "idrac_user": "root",
+                "idrac_password": "password",
+                "share_name": "some_share",
+                "share_user": "some_username",
+                "share_password": "some_password"
+            }
+        }
+
```
Comment (rajeevarakkal) [inline .github/debug.md:24]: Add a link of ansible doc on how inject arguments to verify the modules locally
```suggestion
For more information about injecting arguments , see  the Ansible [docs](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#exercising-module-code-locally).
```

File: .github/debug.md [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+5. Install ansible with `pip install ansible`
```
Comment (rajeevarakkal) [inline .github/debug.md:25]: It is expected that `anisble `installed already and we are building on top of that. You can safely remove this line

```suggestion

```

File: .github/debug.md [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+6. Execute by running `python -m ansible.modules.some_module /tmp/args.json`. How to run this in an IDE is described below.
```
Comment (rajeevarakkal) [inline .github/debug.md:26]:
```suggestion
6. Execute one of the modules by running `python -m ansible.modules.<some_module> /tmp/args.json`.  

How to run this in an IDE is described below.
```

File: .github/debug.md [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+## PyCharm
+
+### Windows Only
```
Comment (rajeevarakkal) [inline .github/debug.md:30]:
```suggestion
### on a remote Windows
```

File: .github/debug.md [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+
+1. Running PyCharm on Windows: The biggest problem with this is that the Ansible Python package does not install cleanly on Windows. The easiest way to fix this is to run a Linux VM and then use a remote debugger. I installed Ubuntu but any version of Linux will work.
+2. Download a copy of [the code](https://github.com/dell/dellemc-openmanage-ansible-modules) and open the folder with PyCharm.
+3. Go to File->Settings->Project:<name_of_your_project>->Python Interpreter
+4. Click the gear and then click add
+5. Use `SSH Interpreter` and then add the Ubuntu box mentioned above or another remote target of your choosing. If you are already on Linux you can use a `Virtualenv environment` or `System Interpreter`
+
+### All Systems
```
Comment (rajeevarakkal) [inline .github/debug.md:38]:
```suggestion
### on a local Linux
```

File: .github/debug.md [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        }
+
+5. Install ansible with `pip install ansible`
+6. Execute by running `python -m ansible.modules.some_module /tmp/args.json`. How to run this in an IDE is described below.
+
+## PyCharm
+
+### Windows Only
+
+1. Running PyCharm on Windows: The biggest problem with this is that the Ansible Python package does not install cleanly on Windows. The easiest way to fix this is to run a Linux VM and then use a remote debugger. I installed Ubuntu but any version of Linux will work.
```
Comment (rajeevarakkal) [inline .github/debug.md:32]: make this as an overview
```suggestion
The following steps helps to run and debug OpenManage Ansible Modules installed on the Linux VM using a remote debugger configured on Windows PyCharm IDE. 
```

File: .github/debug.md [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+2. Download a copy of [the code](https://github.com/dell/dellemc-openmanage-ansible-modules) and open the folder with PyCharm.
```
Comment (rajeevarakkal) [inline .github/debug.md:33]:
```suggestion
1. Download a copy of [the code](https://github.com/dell/dellemc-openmanage-ansible-modules) and open the folder with PyCharm.
```

File: .github/debug.md [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+3. Go to File->Settings->Project:<name_of_your_project>->Python Interpreter
```
Comment (rajeevarakkal) [inline .github/debug.md:34]:
```suggestion
2. Go to File->Settings->Project:<name_of_your_project>->Python Interpreter
```

File: .github/debug.md [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+4. Click the gear and then click add
+5. Use `SSH Interpreter` and then add the Ubuntu box mentioned above or another remote target of your choosing. If you are already on Linux you can use a `Virtualenv environment` or `System Interpreter`
```
Comment (rajeevarakkal) [inline .github/debug.md:36]:
```suggestion
4. Use `SSH Interpreter` and then add the Ubuntu box mentioned above or another remote target of your choosing. If you are already on Linux you can use a `Virtualenv environment` or `System Interpreter`
```

File: .github/debug.md [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+
+### All Systems
+
+1. You will need to configure the IDE to use the `args.json` file you created above. In PyCharm do this by going to Run-Edit Configurations. In `Parameters` add `<ABSOLUTE_PATH>\args.json`. For example: `C:\Users\grant\Documents\dellemc-openmanage-ansible-modules\args.json`. This will pass the JSON file as an argument to the module when it runs. You should now be able to debug the module directly.
+2. It is also possible to pass the arguments within the Python script itself by updating the `main` function with:
+
+        basic._ANSIBLE_ARGS = to_bytes(json.dumps({'ANSIBLE_MODULE_ARGS': {"hostname": "100.100.29.247", "username": "admin", "password": "Dell_123"}}))
+        set_module_args(args)
+3. If you would like to set PYTHONPATH with PyCharm you can do that by going to Run->Edit Configurations->Environment Variables
```
Comment (rajeevarakkal) [inline .github/debug.md:45]: please remove this image file as well. Adding image file is something which is not recommended. It obvious from this statement itself on how to make pythonpath changes.

File: .github/debug.md [hunk 20]
```diff
@@ -0,0 +1,47 @@
...
+
+1. Running PyCharm on Windows: The biggest problem with this is that the Ansible Python package does not install cleanly on Windows. The easiest way to fix this is to run a Linux VM and then use a remote debugger. I installed Ubuntu but any version of Linux will work.
+2. Download a copy of [the code](https://github.com/dell/dellemc-openmanage-ansible-modules) and open the folder with PyCharm.
+3. Go to File->Settings->Project:<name_of_your_project>->Python Interpreter
+4. Click the gear and then click add
+5. Use `SSH Interpreter` and then add the Ubuntu box mentioned above or another remote target of your choosing. If you are already on Linux you can use a `Virtualenv environment` or `System Interpreter`
+
+### All Systems
+
+1. You will need to configure the IDE to use the `args.json` file you created above. In PyCharm do this by going to Run-Edit Configurations. In `Parameters` add `<ABSOLUTE_PATH>\args.json`. For example: `C:\Users\grant\Documents\dellemc-openmanage-ansible-modules\args.json`. This will pass the JSON file as an argument to the module when it runs. You should now be able to debug the module directly.
```
Comment (rajeevarakkal) [inline .github/debug.md:40]:
```suggestion
1. You will need to configure the IDE to use the `args.json` file you created above. In PyCharm do this by going to Run-Edit Configurations. In `Parameters` add `<ABSOLUTE_PATH>\args.json`.  This will pass the JSON file as an argument to the module when it runs. You should now be able to debug the module directly.
```

File: .github/debug.md [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+2. It is also possible to pass the arguments within the Python script itself by updating the `main` function with:
+
+        basic._ANSIBLE_ARGS = to_bytes(json.dumps({'ANSIBLE_MODULE_ARGS': {"hostname": "100.100.29.247", "username": "admin", "password": "Dell_123"}}))
```
Comment (rajeevarakkal) [inline .github/debug.md:43]:
```suggestion
        basic._ANSIBLE_ARGS = to_bytes(json.dumps({'ANSIBLE_MODULE_ARGS': {"hostname": "192.168.0.1", "username": "idrac_user", "password": "idrac_pwd"}}))
```

**Comment by grantcurell** [comment]
All requested changes made. I proof read the updated content for grammar. I detailed my changes in the last commit.

**Review by jagadeeshnv — APPROVED** [review]

**Review by rajeevarakkal — APPROVED** [review]
